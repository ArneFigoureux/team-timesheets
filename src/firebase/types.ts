export enum LoginProvider {
    Email = 0,
    Facebook = 1,
    Google = 2,
    Guest = 3,
}
