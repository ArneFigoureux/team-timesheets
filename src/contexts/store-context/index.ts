export * from "./store-context";
export * from "./store-provider";
export * from "./use-store";
