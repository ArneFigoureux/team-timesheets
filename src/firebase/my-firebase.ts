import "firebase/firestore";
import * as firebase from "firebase/app";
import { config } from "../config";

const app = firebase.initializeApp({
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket
});

export const firestore = app.firestore();
export const storage = app.storage();
export const auth = app.auth();
export const functions = app.functions();