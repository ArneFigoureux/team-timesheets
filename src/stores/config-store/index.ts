export * from "./config-store";
export * from "./use-config-store";