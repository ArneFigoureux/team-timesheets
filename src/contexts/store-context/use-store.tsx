import React from "react";
import { StoreContext } from "./store-context";

export const useStore = () => React.useContext(StoreContext);