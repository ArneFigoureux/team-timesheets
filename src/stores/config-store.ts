import { observable, computed } from 'mobx';
import { Collection, ICollection, RealtimeMode, FetchMode } from "firestorable";
import { IRootStore } from './root-store';
import { firestore } from '../firebase/my-firebase';
import { IProject, ITask, IClient, IClientData, ITeam, ITeamData, ITaskData, IConfig, ConfigValue } from '../../common/dist';

import * as serializer from '../../common/serialization/serializer';
import * as deserializer from '../../common/serialization/deserializer';
import { LoginProvider } from '../firebase/types';

export type Configs = {
    // authClientId: string;
    loginProviders: LoginProvider[];
}

export interface IConfigStore extends ConfigStore {};

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;
    readonly tasks: ICollection<IProject, ITaskData>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam, ITeamData>;
    readonly configsCollection: Collection<IConfig>;

    @observable.ref taskId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore) {
        // this._rootStore = rootStore;
        const deps = { logger: console.log };

        this.tasks = new Collection<ITask, ITaskData>(
            firestore,
            "tasks",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.clientsCollection =
            new Collection<IClient, IClientData>(
                firestore,
                "clients",
                {
                    realtimeMode: RealtimeMode.on,
                    fetchMode: FetchMode.once,
                    query: ref => ref.orderBy("name_insensitive"),
                    serialize: serializer.convertNameWithIcon,
                    deserialize: deserializer.convertNameWithIcon,
                },
                deps,
            );

        this.teamsCollection = new Collection<ITeam, ITeamData>(
            firestore,
            "teams",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.configsCollection = new Collection<IConfig>(
            firestore,
            "configs",
            {
                fetchMode: FetchMode.once,
            }
        )
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }

    // To investigate: does getConfigValue needs mobx @computed attribute?
    public getConfigValue<T>(key: string): T;
    public getConfigValue<T>(key: string, isRequired: false): T | undefined;
    public getConfigValue<T extends ConfigValue>(key: string, isRequired = true): T | undefined {
        const doc = this.configsCollection.docs.find(c => c.data!.key === key);

        if (!doc) {
            if (isRequired) throw new Error(`Required config '${key}' is missing.`);
            else return undefined;
        }

        return doc.data!.value as unknown as T;
    }
}
