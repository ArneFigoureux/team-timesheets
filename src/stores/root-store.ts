import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { firestore } from "../firebase/my-firebase";
import { IProjectStore, ProjectStore } from "./project-store";

export interface IRootStore {

    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly getCollection: (name: string) => firebase.firestore.CollectionReference;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;

    public readonly getCollection: (name: string) => firebase.firestore.CollectionReference;

    constructor(getCollection: (name: string) => firebase.firestore.CollectionReference) {
        this.getCollection = getCollection;

        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this, getCollection);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);

    }
};

const store = (window as any)["store"] = new Store(name => firestore.collection(name));

export default store;

