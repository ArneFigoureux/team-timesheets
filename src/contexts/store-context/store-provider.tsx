import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store } from "../../stores/root-store";

export const StoreProvider = ({ children }: React.PropsWithChildren<{}>) => {
    const firebase = useFirebase();

    const auth = firebase.auth();
    const firestore = firebase.firestore();
    const storage = firebase.storage();

    const store = (window as any)["store"] = new Store({
        auth,
        firestore,
        storage,
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
