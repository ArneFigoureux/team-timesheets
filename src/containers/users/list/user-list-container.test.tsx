import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";

import { initTestFirestore, deleteFirebaseAppsAsync } from "../../../__tests__/utils/firebase";

import { TestCollection } from "../../../__tests__/utils/firestorable/collection";

import { StoreContext } from "../../../contexts/store-context";
import { Store } from "../../../stores/root-store";
import { ITeamData } from "../../../../common";
import { UserList } from ".";

jest.mock("@material/top-app-bar/index", () => ({
    MDCTopAppBar: () => <></>,
}));

jest.mock("@material/icon-button/index", () => ({
    MDCIconButtonToggle: () => <></>,
}));

jest.mock("@material/ripple/index", () => ({
    MDCRipple: () => <></>,
}));

jest.mock("@material/switch/index", () => ({
    MDCSwitch: () => <></>,
}));

const {
    firestore,
    refs: [
        userRef,
        teamRef,
    ],
    clearFirestoreDataAsync
} = initTestFirestore("user-list-test",
    [
        "users",
        "teams",
    ]);

beforeAll(clearFirestoreDataAsync);
afterAll(() => deleteFirebaseAppsAsync());

const userCollection = new TestCollection(firestore, userRef);
const teamCollection = new TestCollection<ITeamData>(firestore, teamRef);

const setupAsync = () => {

    return Promise.all([
        userCollection.addAsync(
            {
                name: "user 1",
                team: "team-1",
                roles: {
                    user: true,
                }
            },
            "user-1",
        ),
        userCollection.addAsync(
            {
                name: "user 2",
                team: "team-0",
                roles: {
                    user: true,
                }
            },
            "user-2",
        ),
        userCollection.addAsync(
            {
                name: "user 3",
                roles: {
                    user: true,
                }
            },
            "user-3",
        ),
        userCollection.addAsync(
            {
                name: "admin 1",
                team: "team-1",
                roles: {
                    admin: true,
                    user: true,
                }
            },
            "admin-1",
        ),
        teamCollection.addAsync(
            {
                name_insensitive: "TEAM 1",
                name: "team 1",
            },
            "team-1",
        ),
    ]);
};

beforeAll(() => setupAsync());

describe("UserListContainer", () => {

    const store = new Store({ firestore });

    it("should render without users", () => {

        const { asFragment } = render(
            <StoreContext.Provider value={store}>
                <UserList />
            </StoreContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    describe("when logged in user is admin", () => {

        beforeAll(() => {
            store.user.setUser({
                uid: "admin-1",
                displayName: "admin 1",
            } as firebase.User);

            return waitFor(() => store.user.isAuthInitialised);
        });

        it("should render users", async () => {
            const { findByText, asFragment } = render(
                <StoreContext.Provider value={store}>
                    <UserList />
                </StoreContext.Provider>
            );

            await waitFor(
                () => findByText("user 1"),
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it("should navigate to user detail when item is clicked", async () => {
            const goTo = jest.fn();
            store.router.goTo = goTo;

            const { findByText } = render(
                <StoreContext.Provider value={store}>
                    <UserList />
                </StoreContext.Provider>
            );

            const userListItemEl = await findByText("user 1");

            expect(goTo).not.toHaveBeenCalled();

            fireEvent.click(userListItemEl);

            expect(goTo).toHaveBeenCalledWith(
                expect.objectContaining({
                    path: "/user/:id"
                }),
                {
                    id: "user-1"
                },
            );
        });

        describe("when user is added in the collection", () => {
            it("should rerender to add the user to the list", async () => {
                const { queryByText, getByText } = render(
                    <StoreContext.Provider value={store}>
                        <UserList />
                    </StoreContext.Provider>
                );

                // user should originally not be in the list
                expect(queryByText("user b")).toBeNull();

                // add user in database
                const userId = await userCollection.addAsync({
                    name: "user b",
                    roles: {
                        user: true,
                    },
                });

                // list should be updated with new user
                await waitFor(() => expect(getByText("user b")));

                // clean up the database by removing the user again
                await userCollection.deleteAsync(userId);
            });
        });

        describe("when user is removed from the collection", () => {
            let userId: string;
            beforeEach(() => userCollection.addAsync({
                name: "user b",
                roles: {
                    user: true,
                },
            }).then((id) => userId = id));

            it("should rerender to removed the user from the list", async () => {
                const { getByText, queryByText } = render(
                    <StoreContext.Provider value={store}>
                        <UserList />
                    </StoreContext.Provider>
                );

                // user should originally be in the list
                expect(getByText("user b")).not.toBeNull();

                // delete user in database
                await userCollection.deleteAsync(userId);

                // list should be updated to remove the user
                await waitFor(() => expect(queryByText("user b")).toBeNull());
            });
        });

        describe("when user is updated in the collection", () => {
            it("should rerender to show the updated user in the list", async () => {
                const { getByText } = render(
                    <StoreContext.Provider value={store}>
                        <UserList />
                    </StoreContext.Provider>
                );

                // user should originally be in the list
                expect(getByText("user 1"));

                // update user in database
                await userCollection.updateAsync({
                    name: "user 1 b",
                }, "user-1");

                // list should be updated with new user name
                await waitFor(() => expect(getByText("user 1 b")));

                // clean up the database by changing user name again
                await userCollection.updateAsync({ name: "user 1" }, "user-1");
            });
        });
    });
});
