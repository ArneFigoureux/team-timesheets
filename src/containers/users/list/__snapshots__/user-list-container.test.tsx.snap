// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListContainer should render without users 1`] = `
<DocumentFragment>
  <div
    class="mdc-list mdc-list--two-line"
  />
</DocumentFragment>
`;

exports[`UserListContainer when logged in user is admin should render users 1`] = `
<DocumentFragment>
  <div
    class="mdc-list mdc-list--two-line"
  >
    <div
      class="mdc-list-item"
      tabindex="0"
    >
      <i
        class="rmwc-icon material-icons mdc-list-item__graphic"
      >
        person_outline
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          admin 1
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          team 1 - Admin, user
        </span>
      </span>
    </div>
    <div
      class="mdc-list-item"
      tabindex="0"
    >
      <i
        class="rmwc-icon material-icons mdc-list-item__graphic"
      >
        person_outline
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          user 1
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          team 1 - User
        </span>
      </span>
    </div>
    <div
      class="mdc-list-item"
      tabindex="0"
    >
      <i
        class="rmwc-icon material-icons mdc-list-item__graphic"
      >
        person_outline
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          user 2
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          User
        </span>
      </span>
    </div>
    <div
      class="mdc-list-item"
      tabindex="0"
    >
      <i
        class="rmwc-icon material-icons mdc-list-item__graphic"
      >
        person_outline
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          user 3
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          User
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
